### ISSUE TITLE
New Module Idea: Normal Distribution Simulation

### ISSUE THREAD

**Issue by ProfessorAmanda** [ProfessorAmanda/econsimulations#295]
Hi all, 

This is just the beginning kernel of an idea, but if Wayne has time, perhaps he can start to mock this up while I scratch my head about all of the features I want to include.

The basic idea is to provide a playground for students to experiment with how the normal distribution describes datasets and determines probabilities. Students especially often struggle with the connection between the value on the axis and the area under the normal curve. For now, let's focus on the normal distribution, but really this could generalize to any continuous probability distribution.

Wayne, you will want to familiarize yourself with the normal distribution and with javascript's functions for drawing normal density functions, for finding the area under the curve given a value on the axis, and for finding a value on the axis given a probability (this would be an inverse normal probability distribution). I believe we use these in various modules already.

To learn about how the normal distribution works conceptually, check out Chapter 4 of OpenIntroStats: 
[openintro-statistics.pdf](https://github.com/ProfessorAmanda/econsimulations/files/9011563/openintro-statistics.pdf)

My sketch in progress:
[Simulation_NormalDistribution.docx](https://github.com/ProfessorAmanda/econsimulations/files/9011506/Simulation_NormalDistribution.docx)

**Comment by waynew99** [comment]
I've made a prototype on shading under the curve, and it's up on `NormalDistribution` branch.
It turned out to be fairly doable and we can surely build more stuff around it!

**Comment by ProfessorAmanda** [comment]
It's so great!! I already love how you can see the effect that changing the mean and standard deviation has on the area under the curve. 

So now we can get serious about this. Fantastic!

Next steps:

- [ ] I think we only need one picture, i.e. combine the mean, standard deviation, and shading steps together as the first step the user performs.
- [ ] Put the choice of >= or <= inside this statement: P (x [selection] [value]). We can experiment with how this looks. I'm not totally sure what to do with the "value" part?
- [ ] The hardest thing will now be to report the area under the curve. Start looking into the normal distribution functions in  javascript.

**Comment by waynew99** [comment]
Got it! Thanks! I'll work on them and let you know when I make progress.

**Comment by waynew99** [comment]
I've made some progress on the module. They can be checked out on branch `NormalDistribution`
1. The two graphs are combined into one.
2. The choices are put in one line inside the statement. However, it looks pretty rough. I'll keep experimenting and improving the aesthetics.
3. The area under the curve is reported. One bug to be fixed: the area rounds up to 100.00% when it's larger than 99.995%. I'll work on fixing this one.

**Comment by ProfessorAmanda** [comment]
I think it looks great! The math seems right, as much as I was able to see from playing around with it. And I actually like how you have written the probability line for now. If you have an idea for a better way to represent it, let me know, but I think it is fine as is for now. Some notes on what I see so far:

- [ ] Just given the way I teach this material, I'd rather have the probabilities reported as a decimal than as a percentage. 
- [ ] Here is some text for the Intro: The Normal Distribution is one of the most important probability distributions, because it describes a wide range of natural phenomena. You can uniquely identify any normal distribution with just its mean and standard deviation. This module demonstrates how the mean and standard deviation determine the probabilities calculated from the normal probability density function. It is also possible to test whether a given dataset follows a normal distribution by using a chi-square goodness-of-fit test. 
- [ ] Here is some text for the menu: The normal distribution can be described entirely by its mean and standard deviation. Many natural phenomena can be described by this distribution, and it is possible to test whether a given dataset follows a normal distribution. 
- [ ] It would be nice for the user to be able to enter numbers in the probabilities statement, rather than just scrolling to them. I'm attaching a screenshot below of the error I got when I clicked the backspace key to try to enter my own number.
![Screen Shot 2022-07-05 at 1 28 37 PM](https://user-images.githubusercontent.com/53241468/177383558-ac029138-d6ad-437c-9eef-03b2d5a47ca3.png)

I am so excited to be able to use this for teaching!

As a next step, I'd like to be able to have the user click a button and "draw" a set of observations from the distribution they have plotted above and show the user a table of those values and a dot plot of those values (similarly to how we plot them in, say, the law of large numbers or central limit theorem modules). The object of this is to further reinforce what is meant by a "distribution." You'll need to look into how javascript can generate a list of random numbers (and crib a little from what previous research assistants have done). This is a nice next step for you to look into, because we will also need this for the test of normality.

Thanks and let me know if you have questions!

**Comment by waynew99** [comment]
Got it! I've fixed the issues listed above, and I'm working on the next step of drawing observations.

**Comment by waynew99** [comment]
The drawing samples part is up on the branch! Please let me know what you think.
(Sorry! That took a bit longer than expected due to a hard-to-debug highchart usage)

**Comment by ProfessorAmanda** [comment]
Hey! Sorry I missed this update. That definitely is working very well. I have a few comments just about formatting, to make it clear to user that the drawing samples step is distinct.

- [ ] Have the user press a button with the text "Experiment with Drawing Samples from This Distribution" to reveal the rest.
- [ ] I think the sample size should be entered rather than a slider option...maybe copy how we've done this in other modules?
- [ ] The "Draw Samples" button should be "Draw a Sample" 

Let me know what you think! I'll get started on the next part.

**Comment by waynew99** [comment]
Of course! It's up on the branch and you can take a look.

**Comment by ProfessorAmanda** [comment]
Thanks! Looks great. 

Oh, one question (I didn't notice this before): What is the rule that determines when an observation will be highlighted in the sample table? I thought it might be whatever was described in the probability rule, but it seems to be "two-tailed." Like, if the rule is P(x>1), I noticed that observations less than -1 were also highlighted in blue. I hadn't even thought of including the highlighted observations, so you could just remove the highlights, or make the highlights agree with the probability rule specified by the user.

**Comment by waynew99** [comment]
Sure. I think the currently highlighted rows are the points that are drawn.

I realized that I was probably doing the samples not the way we wanted – on every "draw a sample" click, a brand new 100 points population is generated, and a desired number of them are selected. I think we want instead is that when mean and standard deviation change, we have a new population of points, and on every "draw a sample" click, we select samples from that same population. Is that correct? If so, should we keep highlighting the rows of the points that are drawn?

**Comment by ProfessorAmanda** [comment]
Hi Wayne, yes exactly, you want to drop a new set of points from a normal distribution with the selected mean and standard deviation. You shouldn't be highlighting points, because they all should be plotted.

**Comment by waynew99** [comment]
Thanks! That makes sense. I've made the changes and pushed onto the branch. I'll get started on the next part.

**Comment by waynew99** [comment]
Hi Amanda, I've pushed a basic prototype including everything up to dividing the histogram into bins. I'll keep working on the rest of it. Please let me know any changes you would like me to make! Thanks!

**Comment by ProfessorAmanda** [comment]
This looks like a good start! 

I have one note on this: before dividing the data into bins, I'd like to show the points as dots, as you did above, so that the user can see the "raw" sample points.

**Comment by waynew99** [comment]
I'm still working on the rest of the module. I've pushed some updates on displaying the raw points and letting the user decide the number of bins. I'm still working on the rest of it where we display the table, as the way Highcharts divides its data into bins is a bit tricky to work with.

**Comment by ProfessorAmanda** [comment]
Very cool! Maybe hold off on displaying the blue bins until the user inputs the number of bins they would like, but otherwise, proceed! I bet the visuals are a little tricky.

**Comment by waynew99** [comment]
It took me some time to fix the little bugs, but I think I got it working! The table is now synced with the histogram, and also shows the expected frequency.

**Comment by ProfessorAmanda** [comment]
Hi Wayne, yes, that's right. The data values should be on the x axis, and the y-axis should represent frequencies of those data values. For a histogram with vertical bars, the x-axis values show you which values of the data points belong in which bins, and the y-axis shows you how many points fit that criterion (if that makes sense).

**Comment by waynew99** [comment]
Got it! Thanks, Amanda! I've made the change and pushed. I'm starting the Chi-squared module.

**Comment by ProfessorAmanda** [comment]
Thanks, Wayne! The axes look right now. I think I now see an error in the generation of the uniform dots. When I plotted a sample with a mean of 0 and standard deviation of 1, I got a range of points from -10 to 10. The formula for the standard deviation of a uniform distribution is sqrt((B-A)^2/12), where A and B are the endpoints. You might want to check which parameters the given javascript function requires.

**Comment by waynew99** [comment]
Currently we are using distribution generation functions from this library: https://statisticsblog.com/probability-distributions/#uniform.
The uniform distribution generation function takes in three parameters: sampleSize, lowerBound, and upperBound. I'm not exactly sure how we can feed standard deviation to this function. Do you think we should explore functions from other libraries instead?

**Comment by ProfessorAmanda** [comment]
I think it should be possible to algebraically back out the lower bound and upper bound from the mean and standard deviation. Let me do some algebra for a few minutes (I'm procrastinating on something scary, lol).

**Comment by ProfessorAmanda** [comment]
Hi Wayne,

Here are expressions for the lower bound and upper bound using the mean and standard deviation:

Lower bound = mean - stddev*sqrt(3)
Upper bound = mean + stddev*sqrt(3)

Give those a shot, and let's see if the picture looks more reasonable.

**Comment by waynew99** [comment]
Ah, got it! Thanks a lot. This makes a lot of sense. The relationship between the standard deviation and the two bounds didn't click in my head for some reason. I'll give this a shot.

**Comment by waynew99** [comment]
Hi Amanda,

It's up on the branch. Looks like the bounds are now changing correctly according to the stddev and mean. Please let me know if it looks good to you!

**Comment by ProfessorAmanda** [comment]
Awesome! Looks much more reasonable. 

- [ ] Is it possible for the size of the graph to adjust based on the standard deviation the user chooses? 
- [ ] Sorry I didn't catch this earlier, but can you check how you are calculating the expected frequencies? It looks like your expected frequencies follow a uniform distribution (same count across all the bins). We'd like expected frequencies based on a normal distribution, which will tend to have more points towards the middle. You will need to figure out the probabilities of being between the endpoints of each bin for a normal distribution, then multiply that by the total sample size, to figure out what number we expect in each bin.

If you tell me what javascript function you have to figure out normal probabilities and what its inputs are, I can help you figure out the probabilities for each bin.

**Comment by waynew99** [comment]
Sure!
1. For the size of the graph, do you mean changing the scale of the x-axis to make sure we don't have a too much "squeezed" chart for a low stadard deviation value?
2. For the expected frequencies:
  - For normal distribution, I'm using the function `nd.cdf(bin.upperBound) - nd.cdf(bin.lowerBound)) * sampleSize`, in which nd.cdf is the library function I'm using to calculate the cumulative distribution function for the normal distribution.
  - For the uniform distribution, I'm simply doing `sampleSize / numberOfBins` for each bin.

**Comment by ProfessorAmanda** [comment]
1. Yes exactly. 
2. Aha! You want to use the normal expected frequencies in *both* cases, since the idea is to test whether the dataset is drawn from a normal distribution. In the case where the dataset is uniform, you'll have that normal cdf in the background (given the mean and standard deviation the user inputs), but you'll never plot it. Eventually, we will hide from the user the information about whether you draw from the uniform or normal distribution to get the dataset.

**Comment by waynew99** [comment]
I see! Thanks! I've pushed the changes.

**Comment by ProfessorAmanda** [comment]
Lookin' good! Proceed!

**Comment by waynew99** [comment]
Hi Amanda,

I spent some time digesting the concepts of hypothesis tests, chi-squared distribution, and p-value. I found this "Chi-square goodness-of-fit test" function from this library: https://stdlib.io/docs/api/latest/@stdlib/stats/chi2gof.
I'm not entirely sure that I'm understanding it correctly, but what I have right now is that I'm feeding the goodness-of-fit test with the observed frequencies for each bin, expected frequencies for each bin, and user defined alpha. The function performs the calculation and returns the result, including the pValue and the test statistic. Depending on whether the pvalue is smaller than the alpha, we either accept or reject the null hypothesis.

The changes are on the branch. Please let me know if any of these is incorrect. Thanks!

**Comment by ProfessorAmanda** [comment]
You've done it!! I really think this is working. I'm going to check the numbers another time, but for, just a few small things.

- [ ] I was really glad to see you implement the option of chi-square as a randomly selected distribution shape. I think it's range looks correct given the inputted mean and standard deviation, but I can check that later. 
- [ ] Okay now for some small text things. There is a typo in the text "Randomly chose [whatever] as distribition shape." (word "distribution").
- [ ] When the user tries to use too many bins, the text should say: "Cannot ensure at least 5 sample points per bin." (not "at least 5 samples")
- [ ] In the conclusion to the hypothesis test, replace "Accept the null hypothesis" with the text "Fail to reject the null hypothesis." The only possible outcomes are "Reject the null hypothesis" or "Fail to reject the null hypothesis."

Thanks again, Wayne! I think this will be ready for deployment and beta testing soon. Amazing!

If we really run out of stuff to do, I have an idea for a crazy simulation demonstrating why this hypothesis test works....but let's table that for now :)

**Comment by waynew99** [comment]
Awesome! Thanks for pointing out the typos. I need to turn on spellcheck in my code editor...

Before we come up with a plan for the next module, I can pick up working on some of the long-term enhancement issues.

**Comment by ProfessorAmanda** [comment]
Sounds great!

**Comment by ProfessorAmanda** [comment]
Maybe we should try to deploy this and ask for beta testers? I can tweet it out to friends.

**Comment by waynew99** [comment]
Sounds good! I will create the PR.

**Comment by ProfessorAmanda** [comment]
Hi Wayne,

In the latest version of the Master branch, I'm getting this error when I try to run the Normal Distribution simulation. I think this is the only one that gives me this error. I ran "npm install" just in case, and that did not fix it. Can you take a look? 
![Screen Shot 2022-08-03 at 10 58 13 AM](https://user-images.githubusercontent.com/53241468/182641118-7b4b7f8f-054b-4004-a765-96846a6909f5.png)

**Comment by waynew99** [comment]
Did you try `npm install --legacy-peer-deps` ?

**Comment by ProfessorAmanda** [comment]
Ah shoot no in my sleepy hurried state I did not! Let me try again in a bit.

Amanda G. Gregg
Associate Professor of Economics, Middlebury College
Join My Personal Zoom Room<https://middlebury.zoom.us/my/agregg?pwd=OWlGMmZMSWJaUkowRG5DUWJtRm9CQT09> (Password: EconHist)
Office: Farrell House 101
Office Phone: (802) 443 - 3419
Pronouns: she/her/hers

________________________________

**Comment by ProfessorAmanda** [comment]
Hi Wayne, can you take a look at the expected frequencies for the case where the underlying data happens to come from a chi-square distribution? I think they might still be calculated incorrectly. Thanks!

**Comment by waynew99** [comment]
Sure. Working on it!

**Comment by waynew99** [comment]
Hi Amanda,
I think I've found the issue. The expected frequencies are always calculated according to the normal distribution function generated using the input data. Just to double check, what we want is that the expected frequencies comes from a chi-square function if the data is chi-square distribution, and all with the same value if the data comes from a uniform distribution. Is that right?

**Comment by ProfessorAmanda** [comment]
Hi Wayne,

No, the expected frequencies should always be calculated using a normal distribution. It looked like, in the chi-square case, the expected frequencies were calculated from a chi-square distribution, but I might have just been eye-balling it wrong.

Amanda G. Gregg
Associate Professor of Economics, Middlebury College
Join My Personal Zoom Room<https://middlebury.zoom.us/my/agregg?pwd=OWlGMmZMSWJaUkowRG5DUWJtRm9CQT09> (Password: EconHist)
Office: Farrell House 101
Office Phone: (802) 443 - 3419
Pronouns: she/her/hers

________________________________

**Comment by waynew99** [comment]
Got it! In that case, since the default mean for the normal distribution is zero, and the chi-squared distribution only starts from zero, it could be that the expected frequencies for the right-half of the normal distribution looking similar to the frequencies of a chi-squared distribution.

**Comment by ProfessorAmanda** [comment]
Aha! Yeah I think that's what's going on. Just to double check: how are you reconciling the parameters of the chi-square with the parameters the user puts in? Given the user can input negative means, there'll be some wonkiness there, since as you point out, the minimum value of chi-square is zero, but that's part of the point of the test anyway.

Amanda G. Gregg
Associate Professor of Economics, Middlebury College
Join My Personal Zoom Room<https://middlebury.zoom.us/my/agregg?pwd=OWlGMmZMSWJaUkowRG5DUWJtRm9CQT09> (Password: EconHist)
Office: Farrell House 101
Office Phone: (802) 443 - 3419
Pronouns: she/her/hers

________________________________

**Comment by waynew99** [comment]
The [function](https://statisticsblog.com/probability-distributions/#chisquared) we are using to generate chi-square distribution takes in two parameters – sample size and degree of freedom. As for now, only "sample size" in the user inputs impacts the chi-square distribution, since we are set on 3 as degree of freedom.

**Comment by ProfessorAmanda** [comment]
Aha! So the three degrees of freedom are relevant when you're getting the p-value from the test. But when you're generating the data in the random sample, your degrees of freedom have to match up to the mean and standard deviation the user enters.

You will not be able to make the data perfectly agree with what the user enters, but we can fudge it slightly. Set the degrees of freedom for the chi-square that generates the data to be 2*the mean if the user enters a mean greater than 0 and, say, .5 otherwise. Give that a shot!

For the chi-square you use at the end to get the p-value, you will still use 3 degrees of freedom.

Amanda G. Gregg
Associate Professor of Economics, Middlebury College
Join My Personal Zoom Room<https://middlebury.zoom.us/my/agregg?pwd=OWlGMmZMSWJaUkowRG5DUWJtRm9CQT09> (Password: EconHist)
Office: Farrell House 101
Office Phone: (802) 443 - 3419
Pronouns: she/her/hers

________________________________

**Comment by waynew99** [comment]
Got it, thanks! I'll make the tweak!

**Comment by ProfessorAmanda** [comment]
Hi Wayne, thanks for making these changes! I have a few more edits:

- [ ] When you show the proportion of points that really are in the specified range of the normal distribution (text box), it should say: "Proportion of Sample Points in Range: 0.460" instead of "Proportion of Samples in Range: 0.460." It's the tiniest thing, but it confuses students.
- [ ] I'm sorry, I think I got the chi-square thing a little wonky, so the resulting expected frequencies are still very strange when the underlying data is actually chi-square. The rule should be the following: Set the degrees of freedom equal to the mean if the mean >= 0 and equal to .5 if the mean < 0.

**Comment by waynew99** [comment]
No problem! I've made the changes, but one scenario looks a bit weird – when the mean equals to zero, the degress of freedom is also zero, resulting in all points having the same value. You can take a look on the `NormalDistribution` branch, and let me know if it's intended behavior. Thanks!

**Comment by ProfessorAmanda** [comment]
You're right. The zero point should be assigned a nonzero degrees of freedom. We should thus make the rule: Set the degrees of freedom equal to the mean if the mean > 0 and equal to .5 if the mean <= 0.

**Comment by waynew99** [comment]
Sounds good! If there's no other changes I can create a PR to merge into `master`.

**Comment by ProfessorAmanda** [comment]
Go for it!

Amanda G. Gregg
Associate Professor of Economics, Middlebury College
Join My Personal Zoom Room<https://middlebury.zoom.us/my/agregg?pwd=OWlGMmZMSWJaUkowRG5DUWJtRm9CQT09> (Password: EconHist)
Office: Farrell House 101
Office Phone: (802) 443 - 3419
Pronouns: she/her/hers

________________________________
